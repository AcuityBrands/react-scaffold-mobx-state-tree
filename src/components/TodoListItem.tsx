/**
 * TodoListItem Compent
 * 
 * Renders LI element with checkbox, label, and prompt when 
 * interacting with the todo list
 */
import * as React from "react";
import { observer } from "mobx-react";
import { ITodoItem } from '../stores/TodoStore'
import { Checkbox, Popconfirm } from 'antd';

interface ITodoListItemProps {
  todo: ITodoItem
}

@observer
export class TodoListItem extends React.Component<ITodoListItemProps, undefined>{
  render() {
    const todo = this.props.todo;
    return (
      <li className={"todo-list-item " + (todo.completed ?"completed":"")}>
        <Checkbox checked={todo.completed} onChange={this.onChange}>{todo.title}</Checkbox>
        <Popconfirm placement="top" title="Delete this todo, are you sure?" onConfirm={this.onRemove} okText="Yes" cancelText="No">
          <a href="javascript:void(0)">remove</a>
        </Popconfirm>
      </li>
    );
  }

  onChange = (e:any) => {
    this.props.todo.changeComplete(e.target.checked);
  }

  onRemove = (e:any) => {
    this.props.todo.remove();
  }
}