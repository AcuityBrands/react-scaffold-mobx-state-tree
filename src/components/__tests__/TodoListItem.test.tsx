import * as React from "react";
import * as Enzyme from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import { TodoListItem } from '../TodoListItem';

Enzyme.configure({ adapter: new Adapter() });

const setup = propOverrides => {
  const props = Object.assign({
    todo: { 
      "userId": 1, "id": 1, "title": "todo1", "completed": true,
      changeComplete: jest.fn(),
      remove: jest.fn()
    }
  }, propOverrides)
  const wrapper = Enzyme.shallow(<TodoListItem {...props} />)
  return {
    props,
    wrapper
  }
}

describe('Todo List Item Component', () => {
  // Minimal component test confirms component rendered
  it("can render", () => {
    const { wrapper } = setup({});
    expect(wrapper.exists()).toBe(true);
  })

  it("matches the snapshot", () => {
    const { wrapper } = setup({});
    expect(wrapper).toMatchSnapshot();
  })

  it("will render a li item", () => {
    const { wrapper } = setup({});
    expect(wrapper.find("li").length).toBe(1);
  })

  it("will attempt to toggle the completed status when changed", () => {
    const { wrapper, props } = setup({});
    const spy = jest.spyOn(props.todo, "changeComplete");
    wrapper.find('Checkbox').simulate('change', {target:{checked:true}});
    expect(spy).toHaveBeenCalled();
    spy.mockClear();
  })

  it("will attempt remove the todo when the remove is confirmed", () => {
    const { wrapper, props } = setup({});
    const spy = jest.spyOn(props.todo, "remove");
    wrapper.find('Popconfirm').simulate('confirm');
    expect(spy).toHaveBeenCalled();
    spy.mockClear();
  })
})