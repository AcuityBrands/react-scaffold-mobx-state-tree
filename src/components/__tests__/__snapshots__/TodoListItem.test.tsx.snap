// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo List Item Component matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TodoListItem
    todo={
      Object {
        "changeComplete": [MockFunction],
        "completed": true,
        "id": 1,
        "remove": [MockFunction],
        "title": "todo1",
        "userId": 1,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Checkbox
          checked={true}
          indeterminate={false}
          onChange={[Function]}
          prefixCls="ant-checkbox"
        >
          todo1
        </Checkbox>,
        <Popconfirm
          cancelText="No"
          okText="Yes"
          okType="primary"
          onConfirm={[Function]}
          placement="top"
          prefixCls="ant-popover"
          title="Delete this todo, are you sure?"
          transitionName="zoom-big"
          trigger="click"
        >
          <a
            href="javascript:void(0)"
          >
            remove
          </a>
        </Popconfirm>,
      ],
      "className": "todo-list-item completed",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "checked": true,
          "children": "todo1",
          "indeterminate": false,
          "onChange": [Function],
          "prefixCls": "ant-checkbox",
        },
        "ref": null,
        "rendered": "todo1",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "cancelText": "No",
          "children": <a
            href="javascript:void(0)"
          >
            remove
          </a>,
          "okText": "Yes",
          "okType": "primary",
          "onConfirm": [Function],
          "placement": "top",
          "prefixCls": "ant-popover",
          "title": "Delete this todo, are you sure?",
          "transitionName": "zoom-big",
          "trigger": "click",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "remove",
            "href": "javascript:void(0)",
          },
          "ref": null,
          "rendered": "remove",
          "type": "a",
        },
        "type": [Function],
      },
    ],
    "type": "li",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Checkbox
            checked={true}
            indeterminate={false}
            onChange={[Function]}
            prefixCls="ant-checkbox"
          >
            todo1
          </Checkbox>,
          <Popconfirm
            cancelText="No"
            okText="Yes"
            okType="primary"
            onConfirm={[Function]}
            placement="top"
            prefixCls="ant-popover"
            title="Delete this todo, are you sure?"
            transitionName="zoom-big"
            trigger="click"
          >
            <a
              href="javascript:void(0)"
            >
              remove
            </a>
          </Popconfirm>,
        ],
        "className": "todo-list-item completed",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "checked": true,
            "children": "todo1",
            "indeterminate": false,
            "onChange": [Function],
            "prefixCls": "ant-checkbox",
          },
          "ref": null,
          "rendered": "todo1",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "cancelText": "No",
            "children": <a
              href="javascript:void(0)"
            >
              remove
            </a>,
            "okText": "Yes",
            "okType": "primary",
            "onConfirm": [Function],
            "placement": "top",
            "prefixCls": "ant-popover",
            "title": "Delete this todo, are you sure?",
            "transitionName": "zoom-big",
            "trigger": "click",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "remove",
              "href": "javascript:void(0)",
            },
            "ref": null,
            "rendered": "remove",
            "type": "a",
          },
          "type": [Function],
        },
      ],
      "type": "li",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
